File: NODE/LinkedLIst.ts
class LinkedListNode {
    value: any
    next: any
    constructor(value) {
        this.value = value;
        this.next = null
    }
}

class LinkedList {
    value: LinkedListNode | null
    head: LinkedListNode | null
    tail: LinkedListNode | null
    size: number;
    constructor(value) {
        let newNode = new LinkedListNode(value);
        this.value = newNode;
        this.head = newNode;
        this.tail = newNode;
        this.size = 1;
    }

    printList() {
        let temp = this.head;
        while(temp !== null) {
            console.log(temp.value);
            temp = temp.next
        }
        console.log("linked list size is ", this.size)
    }

    append(value) {
        let newNode = new LinkedListNode(value);
        if (this.head === null) {
            this.head = newNode;
            this.tail = newNode;
        } else {
            if (this.tail) {
                this.tail.next = newNode;
                this.tail = newNode;
            }
        }
        this.size += 1
    }

    pop() {
        if (this.head === null) return false;
        if (this.size === 1) {
            let temp = this.head
            this.head = null;
            this.tail = null;
            this.size = 0;
            return temp;
        } else {
            let temp = this.head;
            let prev = this.head;
            while(temp.next) {
                prev = temp;
                temp = temp.next
            }
            this.tail = prev
            this.tail.next = null
            this.size -= 1
            return temp
        }
    }

    prepend(value) {
        let newNode = new LinkedListNode(value);
        if (!this.head) {
            this.head = newNode
            this.tail = newNode
        } else {
            newNode.next = this.head
            this.head = newNode
        }
        this.size += 1
    }

    get(index) {
        if (index < 0 || index >= this.size) return false
        else {
            let temp = this.head;
            for(let i = 0; i < index; i++) {
                temp = temp?.next     
            }
            return temp?.value
        }
    }

    set(index, value) {
        if (index < 0 || index >= this.size) return false
        else {
            let temp = this.head;
            for(let i = 0; i < index; i++) {
                temp = temp?.next     
            }
            if (temp?.value) temp.value = value
            return true
        }
    }
}

const myLinkedLIst = new LinkedList(4);
// myLinkedLIst.append(5)
// myLinkedLIst.append(10)
// myLinkedLIst.printList()
console.log(myLinkedLIst.pop())
// myLinkedLIst.printList()
console.log('pop on empty list, show False',myLinkedLIst.pop())
myLinkedLIst.append(20);
// myLinkedLIst.printList();
myLinkedLIst.prepend(99);
// myLinkedLIst.printList()
myLinkedLIst.append(25);
myLinkedLIst.append(50);
myLinkedLIst.printList();
console.log(myLinkedLIst.get(0))
console.log(myLinkedLIst.get(1))
console.log(myLinkedLIst.get(2))
console.log(myLinkedLIst.get(3))
console.log(myLinkedLIst.get(-1))
myLinkedLIst.set(3, 19999);
console.log(myLinkedLIst.get(3))

File: PYTHON/LinkedList.py
class Node:
    def __init__(self, value):
        self.value = value
        self.next = None

class LinkedList:
    def __init__(self, value) -> None:
        new_node = Node(value)
        self.head = new_node
        self.tail = new_node
        self.size = 1

    def print_list(self):
        temp = self.head
        while temp is not None:
            print(temp.value)
            temp = temp.next
        print(self.size)
    
    def append(self, value):
        """ check if LL is empty """
        new_node = Node(value)
        if self.head is None:
            self.head = new_node
            self.tail = new_node
        else:
            self.tail.next = new_node
            self.tail = new_node
        self.size += 1

    def pop(self):
        if self.head is None:
            return False
        
        if self.size is 1:
            self.head = None
            self.tail = None
            self.size = 0
            return True
        else:
            temp = self.head
            prev = self.head
            while temp.next:
                prev = temp
                temp = temp.next
            self.tail = prev
            self.tail.next = None
            self.size -= 1
    
    def prepend(self, value):
        newNode = Node(value)
        if self.head is None:
            self.head = newNode
            self.tail = newNode
        else:
            newNode.next = self.head
            self.head = newNode
        self.size += 1
    
    def popFirst(self):
        if self.head is None:
            return False
        temp = self.head
        if self.size is 1:
            self.head = None
            self.tail = None
        else:
            self.head = self.head.next
            temp.next = None
        self.size -= 1
        return temp

    def get(self, index):
        if index < 0 or index >= self.size:
            return None
        else:
            temp = self.head
            for _ in range(index):
                temp = temp.next
            return temp.value
    
    def set_value(self, index, value):
        if index < 0 or index >= self.size:
            return None
        else:
            temp = self.head
            for _ in range(index):
                temp = temp.next
            temp.value = value
        return True
        
                



    
    

linkedList = LinkedList(4)
linkedList.append(5)
linkedList.append(7)
# linkedList.print_list()
linkedList.pop();
# linkedList.print_list()
linkedList.prepend(99)
# linkedList.print_list()
print(linkedList.popFirst())
# linkedList.print_list()
linkedList.append(55)
linkedList.append(89)
linkedList.print_list()
print(linkedList.get(3))
linkedList.set_value(3, 10000)
linkedList.print_list()